encode_utils/connection.py

```diff
@@ -1118,19 +1118,20 @@ def patch(self, payload, raise_403=True, extend_array_values=True):
             for key in payload:
                 if isinstance(payload[key], list):
                     val = payload[key]
-                    if not isinstance(val, str):
-                        # Currently only able to work with arrays of strings. 
-                        continue
                     val.extend(rec_json.get(key, []))
                     # I use rec_json.get(key,[]) above because in a GET request,
                     # not all props are pulled back when they are empty.
                     # For ex, in a file object, if the controlled_by prop isn't set, then
                     # it won't be in the response.
- 
-                    ## CHECK FOR DUPLICATES: Be careful as some can be tricky, i.e.
-                    # ['/documents/id1', 'id1'] 
-                    # such a duplicate should be identified and removed, leaving us with ["id1"]. 
-                    payload[key] = eup.remove_duplicate_associations(val)
+
+                    if isinstance(val[0], str):
```

Comment:
Is it possible for val to have mixed types? If so, we should check that element of the array is a string.

Reply:
I think the answer is no, but there is sort of an exception for arrays of dictionaries.

I think encode_utils tries to check arrays of dicts in [eu_register.py](https://github.com/StanfordBioinformatics/encode_utils/blob/master/encode_utils/MetaDataRegistration/eu_register.py#L315-L324).
And it does so by attempting to json.loads() the string in the field. However, it looks like json.loads() will parse a list of mixed types correctly according to the [python docs](https://docs.python.org/3/library/json.html). 

So eu_register.py will let you (attempt to) submit non-dictionary values for arrays of dicts. The error only occurs when you try to actually patch an object, at which point you get a failed validation error because the portal detects that the array has something that's not a dictionary. I guess we should edit check_valid_json to detect this problem, otherwise users won't realize something's wrong when doing a dry run, but this seems like a separate ticket.

For arrays of other types, it will cast every value in the array as the item_val_type which is grabbed from the schema, which I think should work for preventing mixed types.
(I guess you could still enter "wrong" things like ['abc',2,{'my':'dict'}] and end up with everything cast as a string, but I'm not sure we should bother catching this.)

Reply:
Thanks for the link. It looks like the types are guaranteed to be consistent between elements of the array, so checking just the first element is fine.